cli/src/main/java/keywhiz/cli/commands/DescribeAction.java

```diff
@@ -83,20 +84,49 @@ public DescribeAction(DescribeActionConfig describeActionConfig, KeywhizClient c
         break;
 
       case "secret":
-        SanitizedSecret sanitizedSecret;
-        try {
-          sanitizedSecret = keywhizClient.getSanitizedSecretByName(name);
-
-          printing.printSanitizedSecretWithDetails(sanitizedSecret);
-        } catch (NotFoundException e) {
-          throw new AssertionError("Secret not found.");
-        } catch (IOException e) {
-          throw Throwables.propagate(e);
-        }
+        describeSecret();
         break;
 
       default:
         throw new IllegalArgumentException("Invalid describe type specified: " + firstType);
     }
   }
+
+  private void describeSecret() {
+    SanitizedSecret sanitizedSecret = getSanitizedSecret();
+    List<SanitizedSecret> deletedSecrets = describeActionConfig.includeDeleted
+        ? getDeletedSecrets()
+        : null;
+
+    if (sanitizedSecret != null) {
+      printing.printSanitizedSecretWithDetails(sanitizedSecret);
+    }
+
+    if (deletedSecrets != null) {
+      printing.printDeletedSecretsWithDetails(deletedSecrets);
+    }
+  }
+
+  @Nullable
+  private SanitizedSecret getSanitizedSecret() {
+    try {
+      return keywhizClient.getSanitizedSecretByName(describeActionConfig.name);
+    } catch (NotFoundException e) {
+      if (!describeActionConfig.includeDeleted) {
+        throw new AssertionError("Secret not found.");
+      }
+      // If we're including deleted secrets, it's ok if no non-deleted secret was found
+      return null;
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private List<SanitizedSecret> getDeletedSecrets() {
+    try {
+      return keywhizClient.getDeletedSecretsByName(describeActionConfig.name);
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
```

Comment:
I don't know if getDeletedSecretsByName() can return a null collection, but if so we could hide that by instead returning an empty collection and eliminate the need for null checks in the calling code.

Reply:
this "everything might be null" thing is so weird to me in a statically typed language

Reply:
I'm going to try your other suggestion of making the lower-level methods (the print methods) handle nulls, so it's fine if this returns null

Reply:
Yeah, the typing in, say, TypeScript where IIRC you can explicitly model the possibility of null via a union type is nice. I think that's one reason why Guava (and Java best practices in general) try to get you to avoid "producing" nulls (whether by replace a null collection with an emtpy one, or with Optional, etc.) so that the code that consumes those values doesn't have to constantly do null checks.